### PR TITLE
Bakerstu tcan4550 driver

### PR DESCRIPTION
**Pull request by bakerstu**

### CODE REVIEW

File: src/freertos_drivers/common/TCAN4550Can.cxx [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+
+                // enable receive timeout interrupt
+                mcanInterruptEnable_.too = 1;
+                register_write(IE, mcanInterruptEnable_.data);
+            }
+        }
+        // shuffle data for structure translation
+        for (size_t i = 0; i < frames_read; ++i)
+        {
+            data[i].can_dlc = mram_rx_buffer[i].dlc;
```
Comment (balazsracz) [inline src/freertos_drivers/common/TCAN4550Can.cxx]: are these aliased onto each other? I'm wondering if this code is breaking the aliasing rules of the compiler, which would make this code subject to breaking under different compilation settings or a newer version of the compiler.
The only meaningful suggestion i have is to compile the target data in a temporary variable, then add a memory barrier instruction `std::atomic_thread_fence()`, then write it out.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 2]
```diff
@@ -0,0 +1,1145 @@
...
+
+        uint16_t address = offset + 0x8000;
+        buf->header.cmd = WRITE;
+        buf->header.addrH = address >> 8;
+        buf->header.addrL = address & 0xFF;
+        buf->header.length = count * (sizeof(MRAMTXBuffer) / 4);
+
+        spi_ioc_transfer xfer;
+        xfer.tx_buf = (unsigned long)&buf->header;
+        xfer.rx_buf = (unsigned long)&buf->header;
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: No, actually not, this is how we get the header status. This should not increase the number of CPU cycles significantly, possibly not at all.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 3]
```diff
@@ -0,0 +1,1145 @@
...
+
+        spi_->transfer_with_cs_assert_polled(&xfer);
+#if TCAN4550_DEBUG
+        HASSERT((buf->header.status & 0x8) == 0);
+#endif
+    }
+
+    void (*interruptEnable_)(); ///< enable interrupt callback
+    void (*interruptDisable_)(); ///< disable interrupt callback
+    unsigned state_ : 4; ///< present bus state
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: Fixed.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 4]
```diff
@@ -0,0 +1,1145 @@
...
+    int spiFd_; ///< SPI bus that accesses MCP2515
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: Fixed.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 5]
```diff
@@ -0,0 +1,1145 @@
...
+    SPI *spi_; ///< pointer to a SPI object instance
+    OSSem sem_; ///< semaphore for posting events
+    MCANInterrupt mcanInterruptEnable_; ///< shaddow for the interrupt enable
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: Fixed.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 6]
```diff
@@ -0,0 +1,1145 @@
...
+    uint32_t txCompleteMask_; ///< shaddow for the transmit complete buffer mask
+    uint32_t txPending_ : 1; ///< waiting on a TX active event
+    uint32_t rxPending_ : 1; ///< waiting on a RX active event
+
+    /// Allocating this buffer here avoids having to put it on the
+    /// TCAN4550Can::write() caller's stack.
+    MRAMTXBufferMultiWrite txBufferMultiWrite_ __attribute__((aligned(8)));
+#if TCAN4550_DEBUG
+    volatile uint32_t regs_[64]; ///< debug copy of MCP2515 registers
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: Fixed.

File: src/freertos_drivers/common/TCAN4550Can.hxx [hunk 7]
```diff
@@ -0,0 +1,1146 @@
...
+
+    /// Timeout counter value register definition
+    struct Tocv
+    {
+        union
+        {
+            uint32_t data; ///< raw word value
+            struct
+            {
+                uint32_t tsc  : 16; ///< timeout counter
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/TCAN4550Can.hxx]: Fixed.

**Pull request merged**